torch/distributed/_state_dict_utils.py

```diff
@@ -457,3 +457,97 @@ def tensor_func(
         return False
 
     return True
+
+
+class _TensorInfo(NamedTuple):
+    size: torch.Size
+    dtype: torch.dtype
+
+
+def _broadcast_tensors(
+    full_state_dict: Dict[str, Any],
+    local_state_dict: Dict[str, Any],
+    keys: List[str],
+    device: torch.device,
+    pg: Optional[dist.ProcessGroup] = None,
+) -> None:
+    tensors = []
+    for key in keys:
+        if dist.get_rank() == 0:
+            full_state = full_state_dict[key]
+            assert isinstance(full_state, torch.Tensor)
+            full_tensor = full_state.detach().to(device)
+        else:
+            tensor_info = full_state_dict[key]
+            full_tensor = torch.empty(
+                size=tensor_info.size,
+                device=device,
+                dtype=tensor_info.dtype,
+            )
+
+        tensors.append(full_tensor)
```

Comment:
With this approach, the entire state dict will have to fit into a single GPU? But that's not guaranteed to work. Shouldn't this broadcast and distribute each tensor individually?

There doesn't seem to be a way currently to efficiently load a full state dict checkpoint into a DTensor model with the new `torch.distributed.checkpoint` APIs.

Reply:
No, that's not true, we don't need to fit the entire state_dict into a single GPU. The full_tensor is not kept after the broadcasting and we broadcast every 10 tensors now.